src/yes/yes.rs

```diff
@@ -82,10 +89,115 @@ fn prepare_buffer<'a>(input: &'a str, _buffer: &'a mut [u8; BUF_SIZE]) -> &'a [u
     input.as_bytes()
 }
 
+#[cfg(not(any(target_os = "linux", target_os = "android")))]
 pub fn exec(bytes: &[u8]) {
     let stdout_raw = io::stdout();
     let mut stdout = stdout_raw.lock();
     loop {
         stdout.write_all(bytes).unwrap();
     }
 }
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+pub fn exec(bytes: &[u8]) {
+    use nix::errno::Errno::{ENOSPC, EPIPE};
+    use std::process::exit;
+
+    match try_splice(bytes) {
+        Err(Sys(err)) => match err {
+            EPIPE => {
+                // Our pipe was interrupted, this happens, for example, when
+                // the shell command `yes | head` is run.
+                exit(0);
+            }
+            ENOSPC => {
+                eprintln!("No space left on disk.");
+                exit(1);
+            }
+            _ => {}
+        },
+        _ => {}
+    }
+
+    // If we reach this point, we should fall back to slow writing.
+    let stdout = io::stdout();
+    let mut stdout = stdout.lock();
+    loop {
+        stdout.write_all(bytes).unwrap();
+    }
+}
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+pub fn try_splice(bytes: &[u8]) -> nix::Result<nix::Error> {
+    use libc::{S_IFIFO, S_IFREG};
+    use nix::errno::Errno::UnknownErrno;
+    use nix::fcntl::{fcntl, splice, vmsplice, FcntlArg, OFlag, SpliceFFlags};
+    // use nix::fcntl::OFlag;
+    use nix::sys::stat::fstat;
+    use nix::sys::uio::IoVec;
+    use nix::unistd::pipe;
+    use std::os::unix::io::AsRawFd;
+
+    let stdout = io::stdout();
+    let stdout_stat = fstat(stdout.as_raw_fd())?;
+    let stdout_is_regular = (stdout_stat.st_mode & S_IFREG) != 0;
+
+    if stdout_is_regular {
+        let byte_iovec = &[IoVec::from_slice(&bytes[..])];
+        let (pipe_rd, pipe_wr) = pipe()?;
+
+        let stdout_access_mode = fcntl(stdout.as_raw_fd(), FcntlArg::F_GETFL)?;
+        // Here I'm using OFlag::from_bits_truncate(), instead of OFlag::from_bits(),
+        // because the latter panics for some reason.
+        let mut stdout_oflags = OFlag::from_bits_truncate(stdout_access_mode);
+        let stdout_is_append = stdout_oflags.contains(OFlag::O_APPEND);
+
+        if stdout_is_append {
+            // First we disable append mode, else splice() will return
+            // Sys(EINVAL) error.
+            stdout_oflags.remove(OFlag::O_APPEND);
+            fcntl(stdout.as_raw_fd(), FcntlArg::F_SETFL(stdout_oflags))?;
+            // Here we splice with an output offset that equals the length of
+            // the file. This has effect of appending to the file. Note that
+            // the offset is incremented automatically by the splice()
+            // function.
+            let mut length_offset = stdout_stat.st_size;
```

Comment:
Writing with append will never overwrite what another process writes (it is an ATOMIC seek to the end before every write call). 
```bash
: > hi; yes >> file & pid=$!; sleep 0.1; echo "hello" >> file; sleep 0.1; kill $pid; grep "hello" file
```
should always output "hello". This implementation looks like it will instead overwrite what any other process has written.

Reply:
I see. What do you suggest?

Reply:
I'm not sure there's a possible way to do this correctly the faster way if append mode is set: both `splice` and `sendfile` don't support outputting to a file with the append flag set.